### DISCUSSION TITLE
Irregular objects selection

### DISCUSSION THREAD

**Discussion by wino45** [TomWright/dasel#325] [Q&A] [answered]
Hi

I am trying to select all objects within an array, but some objects does not have all attributes. 
` echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "wrong_name": "Jim"
    }
  ]
}' | dasel -r json 'users.all().name'
` 

I would expect that this simple query returns first array element, while it returns nothing. Can I filter based on existence of an attribute ?

**Comment by TomWright** [comment, marked as answer]
There is something strange going on with the error output that I've been meaning to look into, but this would be throwing an error when it see's an object without the name.

You can get this working by adding a `?` after the "optional" property name, e.g:

```bash
echo '{ "users": [ { "name": "Tom" }, { "wrong_name": "Jim" } ] }' | dasel -r json 'users.all().name?'
"Tom"
```

This is documented here: https://daseldocs.tomwright.me/functions/property
> If any argument ends with a ?, that lookup is considered optional and dasel will continue if that property is not found.

**Comment by wino45** [comment]
Yeah, lack of error is misleading (it was Dasel 2.2).
When reading the property documentation it is not obvious that using ? can be used as filter. Adding one example can help.

```
echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "wrong_name": "Jim"
    }
  ]
}' | dasel -r json 'users.all().filter(name?)'
{
  "name": "Tom"
}
```

**Reply by TomWright** [reply]
The `?` operator isn't necessarily used as a filter although I can see how it appears that way.
Rather the `filter` func is using the string it finds there and evaluating it as truthy, whilst also skipping any items that returned a not found error.

This is clearer in this example:
```bash
echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": "true"
    },
    {
      "wrong_name": "Jim"
    }
  ]
}' | dasel -r json 'users.all().filter(name?)'
{
  "name": "Tom"
}
{
  "name": "true"
}
```

https://github.com/TomWright/dasel/blob/7b3a7823e8659302fcc6aec088126215d91823f7/func_filter.go#L16-L31

**Reply by TomWright** [reply]
As a side note, I've just fixed the error display issue. Releasing `v2.3.3` with this fix now.

```
$ echo '{ "users": [ { "name": "Tom" }, { "wrong_name": "Jim" } ] }' | dasel -r json 'users.all().name' 
Error: could not access map index: property not found: name
exit status 1
```

**Comment by wino45** [comment]
So the proper object selection is:

```
echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    }
  ]
}' | dasel -r json 'users.all().name?.parent()'
{
  "name": "Tom"
}
{
  "name": "false"
}
{
  "name": true
}
```

I would not find this without your help. Thank you.

**Comment by wino45** [comment]
Any idea how to reverse the search (find objects that does not have a property) ?

This almost work :
```
echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    }
  ]
}' | dasel -r json 'users.all().filter(not(name?))'
{
  "name": "false"
}
{
  "wrong_name": "Jim"
}
```

**Comment by TomWright** [comment]
Note that some of the following examples will only work with dasel v2.3.4 and up due to some bugs that have been fixed.

This shows you users that have a name property that is not blank.
```bash
$ echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    },
    {
      "name": ""
    }
  ]
}' | dasel -r json 'users.all().filter(name?.len())'
{
  "name": "Tom"
}
{
  "name": "false"
}
{
  "name": true
}
```

This shows you users that have no name field, or a blank name field.
```bash
$ echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    },
    {
      "name": ""
    }
  ]
}' | dasel -r json 'users.all().filter(not(name?.len()))'
{
  "wrong_name": "Jim"
}
{
  "name": ""
}
```

This shows you users that have a name field, regardless of whether it is blank or not.
```bash
$ echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    },
    {
      "name": ""
    }
  ]
}' | dasel -r json 'users.all().filter(keys().all().filter(equal(.,name)))'
{
  "name": "Tom"
}
{
  "name": "false"
}
{
  "name": true
}
{
  "name": ""
}
```

This shows you users that do NOT have a name field, regardless of whether it is blank or not.
```bash
$ echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    },
    {
      "name": ""
    }
  ]
}' | dasel -r json 'users.all().filter(not(keys().all().filter(equal(.,name)).len()))'
{
  "wrong_name": "Jim"
}
```

**Comment by wino45** [comment]
Why `len()` is needed in the last example? It works without it:
```
$ echo '{
  "users": [
    {
      "name": "Tom"
    },
    {
      "name": "false"
    },
    {
      "name": true
    },
    {
      "wrong_name": "Jim"
    },
    {
      "name": ""
    }
  ]
}' | dasel -r json 'users.all().filter(not(keys().all().filter(equal(.,name))))'
{
  "wrong_name": "Jim"
}
```

What about defining a new function "isKey(x)" = keys().all().filter(equal(.,x)), so the example will become:
`dasel -r json 'users.all().filter(not(isKey(name)))'`
I would be very easy to define any combination of logic based on existence of a key, example:
`dasel -r json 'users.all().filter(not(or(isKey(name),isKey(wrong_name))))'`


BTW. The [Equal](https://daseldocs.tomwright.me/functions/equal) documentation page uses direct Go invocation in the example.
